crates/wasmtime/src/runtime/vm/cow.rs

```diff
@@ -817,9 +817,33 @@ mod test {
         }
     }
 
-    fn mmap_4mib_inaccessible() -> Mmap<AlignedLength> {
+    fn mmap_4mib_inaccessible() -> Arc<Mmap<AlignedLength>> {
         let four_mib = HostAlignedByteCount::new(4 << 20).expect("4 MiB is page aligned");
-        Mmap::accessible_reserved(HostAlignedByteCount::ZERO, four_mib).unwrap()
+        Arc::new(Mmap::accessible_reserved(HostAlignedByteCount::ZERO, four_mib).unwrap())
+    }
+
+    /// Presents a part of an mmap as a mutable slice within a callback.
+    ///
+    /// The callback ensures that the reference no longer lives after the
+    /// function is done.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that during this function call, the only way this
+    /// region of memory is not accessed by (read from or written to) is via the
+    /// reference. Making the callback `'static` goes some way towards ensuring
+    /// that, but it's still possible to squirrel away a reference into global
+    /// state. So don't do that.
+    unsafe fn with_slice_mut(
```

Comment:
This was somewhat annoying but I think this approach makes sense -- basically ensure that any `&mut` references are very short-lived as required by Rust's [aliasing rules](https://doc.rust-lang.org/std/ptr/index.html#pointer-to-reference-conversion).